File: python/src/aoc_2023_03.py
"""Advent of Code 2023 - Day 3."""

from dataclasses import dataclass
from doctest import testmod
from sys import stdin
from typing import Iterator

from coord import Coord, Extent


@dataclass
class Number:
    """Represent a number on the grid"""

    value: int
    position: Coord  # position of first digit
    length: int  # number of digits


class Schematic:
    """Main class to hold schematic"""

    def __init__(self, lines: list[str]) -> None:
        self.extent: Extent = Extent(x=len(lines[0]), y=len(lines))
        self.data: dict[Coord, str] = {}
        for y, line in enumerate(lines):
            for x, c in enumerate(line):
                self.data[Coord(x, y)] = c

    def numbers(self) -> Iterator[Number]:
        """Yield all the numbers."""
        n: Number | None = None
        for y in range(self.extent.y):
            for x in range(self.extent.x + 1):
                if x == self.extent.x:
                    if n is not None:
                        yield n
                    n = None
                else:
                    c = Coord(x, y)
                    d = self.data[c]
                    if d.isdigit():
                        if n is None:
                            n = Number(value=int(d), position=c, length=1)
                        else:
                            n.value = n.value * 10 + int(d)
                            n.length += 1
                    else:
                        if n is not None:
                            yield n
                        n = None

    def is_symbol(self, c: Coord) -> bool:
        """Is the given coordinate a symbol"""
        return not (self.data[c].isdigit() or self.data[c] == ".")

    def is_symbol_adjacent(self, n: Number) -> bool:
        """Is the number adjacent to a symbol on the grid?"""
        if n.position.y > 0:
            for x in range(
                max(0, n.position.x - 1),
                min(self.extent.x, n.position.x + n.length + 1),
            ):
                if self.is_symbol(Coord(x, n.position.y - 1)):
                    return True
        if n.position.y < self.extent.y - 1:
            for x in range(
                max(0, n.position.x - 1),
                min(self.extent.x, n.position.x + n.length + 1),
            ):
                if self.is_symbol(Coord(x, n.position.y + 1)):
                    return True
        if n.position.x > 0 and self.is_symbol(Coord(n.position.x - 1, n.position.y)):
            return True
        if n.position.x + n.length < self.extent.x and self.is_symbol(
            Coord(n.position.x + n.length, n.position.y)
        ):
            return True
        return False


if __name__ == "__main__":
    testmod()
    schematic = Schematic([s.strip() for s in stdin.readlines()])
    print(sum(n.value for n in schematic.numbers() if schematic.is_symbol_adjacent(n)))

File: rust/src/bin/aoc_2023_03.rs
// Advent of Code, 2023 day 03

use aoc_rust::UCoord;
use std::io::stdin;

struct Number {
    value: u32,
    position: UCoord, // position of first digit
    length: usize,    // number of digits
}

// subtract one unless value is zero
fn dec(x: usize) -> usize {
    if x > 0 {
        x - 1
    } else {
        x
    }
}

impl Number {
    // Is the number adjacent to the coordinate
    fn is_adjacent(&self, c: UCoord) -> bool {
        c.col >= dec(self.position.col)
            && c.col <= self.position.col + self.length
            && c.row >= dec(self.position.row)
            && c.row <= self.position.row + 1
    }

    // Is the number adjacent to any of the coordinates
    fn is_any_adjacent(&self, cs: &[UCoord]) -> bool {
        cs.iter().any(|c| self.is_adjacent(*c))
    }
}

fn gear_ratios(stars: &[UCoord], numbers: &[Number]) -> u32 {
    let mut sum = 0;
    for star in stars {
        let adjacent_numbers: Vec<&Number> =
            numbers.iter().filter(|n| n.is_adjacent(*star)).collect();
        match adjacent_numbers[..] {
            [n1, n2] => {
                sum += n1.value * n2.value;
            }
            _ => {}
        }
    }
    sum
}

fn read_schematic() -> (Vec<Number>, Vec<UCoord>, Vec<UCoord>) {
    let mut numbers: Vec<Number> = Vec::new();
    let mut symbols: Vec<UCoord> = Vec::new();
    let mut stars: Vec<UCoord> = Vec::new();
    let mut current_number: Option<Number> = None;

    for (row, line) in stdin().lines().enumerate() {
        for (col, ch) in line.unwrap().chars().enumerate() {
            if let Some(ch_value) = ch.to_digit(10) {
                current_number = current_number.map_or(
                    Some(Number {
                        value: ch_value,
                        position: UCoord { row, col },
                        length: 1,
                    }),
                    |n| {
                        Some(Number {
                            value: n.value * 10 + ch_value,
                            length: n.length + 1,
                            ..n
                        })
                    },
                )
            } else {
                if let Some(n) = current_number {
                    numbers.push(n);
                    current_number = None;
                }
                if ch != '.' {
                    symbols.push(UCoord { row, col });
                    if ch == '*' {
                        stars.push(UCoord { row, col });
                    }
                }
            }
        }
        if let Some(n) = current_number {
            numbers.push(n);
            current_number = None;
        }
    }
    (numbers, symbols, stars)
}

fn main() {
    let (numbers, symbols, stars) = read_schematic();

    let part_a: u32 = numbers
        .iter()
        .filter(|n| (*n).is_any_adjacent(&symbols))
        .map(|n| n.value)
        .sum();

    println!("{}", part_a);
    println!("{}", gear_ratios(&stars, &numbers));
}
